### ISSUE TITLE
Disallow mappings in memory and copying mappings

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#7739]
Related to #6444 and #6282

Due to the fact that mappings are silently skipped, we should disallow the following operations on any data that contains a mapping:
 - copy from storage (to storage or memory)
 - declare as memory type
 - declare as calldata type

(delete part moved to https://github.com/ethereum/solidity/issues/8535 )

all code generator parts that currently skip mapping should be changed to assertions

**Comment by erak** [comment]
There's a PR (https://github.com/ethereum/solidity/pull/6282) already that partially implements this.

**Comment by hrkrshnn** [comment]
Regarding "declare as calldata type"

Currently, in develop, the following contract:
```solidity
contract test {
    function f(mapping(uint => uint) calldata b) external {
    }
}
```

would produce:

```
Error: Mapping types can only have a data location of "storage" and thus only be parameters or return variables for internal or library functions.
 --> /tmp/calldata.sol:2:16:
  |
2 |     function f(mapping(uint => uint) calldata b) external {
  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Is this what you wanted?

**Comment by chriseth** [comment]
Yes

**Comment by chriseth** [comment]
Closed by #9146
